// @heroui/tabs/dist/chunk-RRHVXFLZ.mjs
"use client";
import {
  tab_panel_default
} from "./chunk-XMFPUKBD.mjs";
import {
  tab_default
} from "./chunk-SK5NDKS3.mjs";
import {
  useTabs
} from "./chunk-CTB32YMK.mjs";

// src/tabs.tsx
import { useEffect, useCallback, useRef, useMemo } from "react";
import { forwardRef } from "@heroui/system";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var Tabs = forwardRef(function Tabs2(props, ref) {
  const {
    Component,
    values,
    state,
    domRef,
    destroyInactiveTabPanel,
    getBaseProps,
    getTabListProps,
    getWrapperProps,
    getTabCursorProps
  } = useTabs({
    ...props,
    ref
  });
  const tabsProps = {
    state,
    listRef: values.listRef,
    slots: values.slots,
    classNames: values.classNames,
    isDisabled: values.isDisabled,
    shouldSelectOnPressUp: values.shouldSelectOnPressUp
  };
  const tabs = [...state.collection].map((item) => /* @__PURE__ */ jsx(tab_default, { item, ...tabsProps, ...item.props }, item.key));
  const variant = props == null ? void 0 : props.variant;
  const previousVariant = useRef(void 0);
  const isVertical = props == null ? void 0 : props.isVertical;
  const previousIsVertical = useRef(void 0);
  const placement = props == null ? void 0 : props.placement;
  const previousPlacement = useRef(void 0);
  const cursorRef = useRef(null);
  const selectedItem = state.selectedItem;
  const selectedKey = selectedItem == null ? void 0 : selectedItem.key;
  const getCursorStyles = useCallback(
    (tabRect) => {
      if (variant === "underlined") {
        return {
          left: `${tabRect.left + tabRect.width * 0.1}px`,
          top: `${tabRect.top + tabRect.height - 2}px`,
          width: `${tabRect.width * 0.8}px`,
          height: ""
        };
      }
      return {
        left: `${tabRect.left}px`,
        top: `${tabRect.top}px`,
        width: `${tabRect.width}px`,
        height: `${tabRect.height}px`
      };
    },
    [variant]
  );
  const withAnimationReset = useCallback(
    (callback) => {
      var _a;
      if (variant !== previousVariant.current || isVertical !== previousIsVertical.current || placement !== previousPlacement.current) {
        (_a = cursorRef.current) == null ? void 0 : _a.removeAttribute("data-animated");
      }
      callback();
      previousVariant.current = variant;
      previousIsVertical.current = isVertical;
      previousPlacement.current = placement;
      requestAnimationFrame(() => {
        var _a2, _b;
        (_a2 = cursorRef.current) == null ? void 0 : _a2.setAttribute("data-animated", "true");
        (_b = cursorRef.current) == null ? void 0 : _b.setAttribute("data-initialized", "true");
      });
    },
    [isVertical, variant, placement]
  );
  const updateCursorPosition = useCallback(
    (selectedTab) => {
      if (!cursorRef.current) return;
      const tabRect = {
        width: selectedTab.offsetWidth,
        height: selectedTab.offsetHeight,
        left: selectedTab.offsetLeft,
        top: selectedTab.offsetTop
      };
      const styles = getCursorStyles(tabRect);
      withAnimationReset(() => {
        cursorRef.current.style.left = styles.left;
        cursorRef.current.style.top = styles.top;
        cursorRef.current.style.width = styles.width;
        cursorRef.current.style.height = styles.height;
      });
    },
    [cursorRef.current, getCursorStyles, withAnimationReset]
  );
  const onResize = useCallback(
    (entries) => {
      const contentRect = entries[0].contentRect;
      if (contentRect.width === 0 && contentRect.height === 0) return;
      updateCursorPosition(entries[0].target);
    },
    [updateCursorPosition]
  );
  useEffect(() => {
    var _a;
    const selectedTab = (_a = domRef.current) == null ? void 0 : _a.querySelector(`[data-key="${selectedKey}"]`);
    if (!selectedTab) return;
    const observer = new ResizeObserver(onResize);
    observer.observe(selectedTab);
    return () => observer.disconnect();
  }, [domRef.current, onResize, selectedKey]);
  const renderTabs = useMemo(
    () => /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("div", { ...getBaseProps(), children: /* @__PURE__ */ jsxs(Component, { ...getTabListProps(), children: [
        !values.disableAnimation && !values.disableCursorAnimation && selectedKey != null && /* @__PURE__ */ jsx("span", { ...getTabCursorProps(), ref: cursorRef }),
        tabs
      ] }) }),
      [...state.collection].map((item) => {
        return /* @__PURE__ */ jsx(
          tab_panel_default,
          {
            classNames: values.classNames,
            destroyInactiveTabPanel,
            slots: values.slots,
            state: values.state,
            tabKey: item.key
          },
          item.key
        );
      })
    ] }),
    [
      Component,
      getBaseProps,
      getTabListProps,
      getTabCursorProps,
      tabs,
      selectedKey,
      state.collection,
      values.disableAnimation,
      values.disableCursorAnimation,
      values.classNames,
      values.slots,
      values.state,
      destroyInactiveTabPanel,
      domRef,
      cursorRef,
      variant,
      isVertical
    ]
  );
  if ("placement" in props || "isVertical" in props) {
    return /* @__PURE__ */ jsx("div", { ...getWrapperProps(), children: renderTabs });
  }
  return renderTabs;
});
var tabs_default = Tabs;

export {
  tabs_default
};
