// @heroui/tabs/dist/tabs.js
"use client";
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/tabs.tsx
var tabs_exports = {};
__export(tabs_exports, {
  default: () => tabs_default
});
module.exports = __toCommonJS(tabs_exports);
var import_react2 = require("react");
var import_system4 = require("@heroui/system");

// src/use-tabs.ts
var import_system = require("@heroui/system");
var import_theme = require("@heroui/theme");
var import_react_utils = require("@heroui/react-utils");
var import_shared_utils = require("@heroui/shared-utils");
var import_react_utils2 = require("@heroui/react-utils");
var import_react = require("react");
var import_tabs = require("@react-stately/tabs");
var import_tabs2 = require("@react-aria/tabs");
function useTabs(originalProps) {
  var _a, _b, _c;
  const globalContext = (0, import_system.useProviderContext)();
  const [props, variantProps] = (0, import_system.mapPropsVariants)(originalProps, import_theme.tabs.variantKeys);
  const {
    ref,
    as,
    className,
    classNames,
    children,
    disableCursorAnimation,
    isVertical = false,
    shouldSelectOnPressUp = true,
    destroyInactiveTabPanel = true,
    ...otherProps
  } = props;
  const Component = as || "div";
  const shouldFilterDOMProps = typeof Component === "string";
  const domRef = (0, import_react_utils.useDOMRef)(ref);
  const disableAnimation = (_b = (_a = originalProps == null ? void 0 : originalProps.disableAnimation) != null ? _a : globalContext == null ? void 0 : globalContext.disableAnimation) != null ? _b : false;
  const placement = (_c = variantProps.placement) != null ? _c : isVertical ? "start" : "top";
  const orientation = isVertical || placement === "start" || placement === "end" ? "vertical" : "horizontal";
  const state = (0, import_tabs.useTabListState)({
    children,
    ...otherProps
  });
  const { tabListProps } = (0, import_tabs2.useTabList)(
    { ...otherProps, orientation },
    state,
    domRef
  );
  const slots = (0, import_react.useMemo)(
    () => (0, import_theme.tabs)({
      ...variantProps,
      disableAnimation,
      ...isVertical ? { placement: "start" } : {}
    }),
    [(0, import_shared_utils.objectToDeps)(variantProps), disableAnimation, isVertical]
  );
  const baseStyles = (0, import_theme.cn)(classNames == null ? void 0 : classNames.base, className);
  const values = (0, import_react.useMemo)(
    () => ({
      state,
      slots,
      classNames,
      disableAnimation,
      listRef: domRef,
      shouldSelectOnPressUp,
      disableCursorAnimation,
      isDisabled: originalProps == null ? void 0 : originalProps.isDisabled
    }),
    [
      state,
      slots,
      domRef,
      disableAnimation,
      disableCursorAnimation,
      shouldSelectOnPressUp,
      originalProps == null ? void 0 : originalProps.isDisabled,
      classNames
    ]
  );
  const getBaseProps = (0, import_react.useCallback)(
    (props2) => ({
      "data-slot": "base",
      className: slots.base({ class: (0, import_theme.cn)(baseStyles, props2 == null ? void 0 : props2.className) }),
      ...(0, import_shared_utils.mergeProps)(
        (0, import_react_utils2.filterDOMProps)(otherProps, {
          enabled: shouldFilterDOMProps
        }),
        props2
      )
    }),
    [baseStyles, otherProps, slots]
  );
  const getWrapperProps = (0, import_react.useCallback)(
    (props2) => ({
      "data-slot": "tabWrapper",
      className: slots.tabWrapper({ class: (0, import_theme.cn)(classNames == null ? void 0 : classNames.tabWrapper, props2 == null ? void 0 : props2.className) }),
      "data-placement": placement,
      "data-vertical": isVertical || placement === "start" || placement === "end" ? "vertical" : "horizontal"
    }),
    [classNames, slots, placement, isVertical]
  );
  const getTabListProps = (0, import_react.useCallback)(
    (props2) => ({
      ref: domRef,
      "data-slot": "tabList",
      className: slots.tabList({ class: (0, import_theme.cn)(classNames == null ? void 0 : classNames.tabList, props2 == null ? void 0 : props2.className) }),
      ...(0, import_shared_utils.mergeProps)(tabListProps, props2)
    }),
    [domRef, tabListProps, classNames, slots]
  );
  const getTabCursorProps = (0, import_react.useCallback)(
    (props2) => ({
      "data-slot": "cursor",
      className: slots.cursor({
        class: (0, import_theme.cn)(classNames == null ? void 0 : classNames.cursor, props2 == null ? void 0 : props2.className)
      })
    }),
    [classNames, slots]
  );
  return {
    Component,
    domRef,
    state,
    values,
    destroyInactiveTabPanel,
    getBaseProps,
    getTabListProps,
    getWrapperProps,
    getTabCursorProps
  };
}

// src/tab.tsx
var import_system2 = require("@heroui/system");
var import_react_utils3 = require("@heroui/react-utils");
var import_shared_utils2 = require("@heroui/shared-utils");
var import_scroll_into_view_if_needed = __toESM(require("scroll-into-view-if-needed"));
var import_focus = require("@react-aria/focus");
var import_tabs3 = require("@react-aria/tabs");
var import_interactions = require("@react-aria/interactions");
var import_theme2 = require("@heroui/theme");
var import_jsx_runtime = require("react/jsx-runtime");
var Tab = (0, import_system2.forwardRef)((props, ref) => {
  var _a;
  const {
    className,
    as,
    item,
    state,
    classNames,
    isDisabled: isDisabledProp,
    listRef,
    slots,
    shouldSelectOnPressUp,
    tabRef,
    ...otherProps
  } = props;
  const { key } = item;
  const domRef = (0, import_react_utils3.useDOMRef)(ref);
  const Component = as || (props.href ? "a" : "button");
  const shouldFilterDOMProps = typeof Component === "string";
  const {
    tabProps,
    isSelected,
    isDisabled: isDisabledItem,
    isPressed
  } = (0, import_tabs3.useTab)({ key, isDisabled: isDisabledProp, shouldSelectOnPressUp }, state, domRef);
  if (props.children == null) {
    delete tabProps["aria-controls"];
  }
  const isDisabled = isDisabledProp || isDisabledItem;
  const { focusProps, isFocused, isFocusVisible } = (0, import_focus.useFocusRing)();
  const { hoverProps, isHovered } = (0, import_interactions.useHover)({
    isDisabled
  });
  const tabStyles = (0, import_theme2.cn)(classNames == null ? void 0 : classNames.tab, className);
  const handleClick = () => {
    if (!(domRef == null ? void 0 : domRef.current) || !(listRef == null ? void 0 : listRef.current)) return;
    (0, import_scroll_into_view_if_needed.default)(domRef.current, {
      scrollMode: "if-needed",
      behavior: "smooth",
      block: "end",
      inline: "end",
      boundary: listRef == null ? void 0 : listRef.current
    });
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    Component,
    {
      ref: (0, import_react_utils3.mergeRefs)(domRef, tabRef),
      "data-disabled": (0, import_shared_utils2.dataAttr)(isDisabledItem),
      "data-focus": (0, import_shared_utils2.dataAttr)(isFocused),
      "data-focus-visible": (0, import_shared_utils2.dataAttr)(isFocusVisible),
      "data-hover": (0, import_shared_utils2.dataAttr)(isHovered),
      "data-hover-unselected": (0, import_shared_utils2.dataAttr)((isHovered || isPressed) && !isSelected),
      "data-key": key,
      "data-pressed": (0, import_shared_utils2.dataAttr)(isPressed),
      "data-selected": (0, import_shared_utils2.dataAttr)(isSelected),
      "data-slot": "tab",
      ...(0, import_shared_utils2.mergeProps)(
        tabProps,
        !isDisabled ? {
          ...focusProps,
          ...hoverProps
        } : {},
        (0, import_react_utils3.filterDOMProps)(otherProps, {
          enabled: shouldFilterDOMProps,
          omitPropNames: /* @__PURE__ */ new Set(["title"]),
          // onClick is now from `tabProps`.
          // omit it to avoid executing onClick it twice.
          omitEventNames: /* @__PURE__ */ new Set(["onClick"])
        }),
        { onClick: (0, import_shared_utils2.chain)(handleClick, tabProps.onClick) }
      ),
      className: (_a = slots.tab) == null ? void 0 : _a.call(slots, { class: tabStyles }),
      title: otherProps == null ? void 0 : otherProps.titleValue,
      type: Component === "button" ? "button" : void 0,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          className: slots.tabContent({
            class: classNames == null ? void 0 : classNames.tabContent
          }),
          "data-slot": "tabContent",
          children: item.rendered
        }
      )
    }
  );
});
Tab.displayName = "HeroUI.Tab";
var tab_default = Tab;

// src/tab-panel.tsx
var import_system3 = require("@heroui/system");
var import_react_utils4 = require("@heroui/react-utils");
var import_shared_utils3 = require("@heroui/shared-utils");
var import_tabs4 = require("@react-aria/tabs");
var import_focus2 = require("@react-aria/focus");
var import_theme3 = require("@heroui/theme");
var import_jsx_runtime2 = require("react/jsx-runtime");
var TabPanel = (0, import_system3.forwardRef)((props, ref) => {
  var _a, _b;
  const { as, tabKey, destroyInactiveTabPanel, state, className, slots, classNames, ...otherProps } = props;
  const Component = as || "div";
  const domRef = (0, import_react_utils4.useDOMRef)(ref);
  const { tabPanelProps } = (0, import_tabs4.useTabPanel)({ ...props, id: String(tabKey) }, state, domRef);
  const { focusProps, isFocused, isFocusVisible } = (0, import_focus2.useFocusRing)();
  const selectedItem = state.selectedItem;
  const content = state.collection.getItem(tabKey).props.children;
  const tabPanelStyles = (0, import_theme3.cn)(classNames == null ? void 0 : classNames.panel, className, (_a = selectedItem == null ? void 0 : selectedItem.props) == null ? void 0 : _a.className);
  const isSelected = tabKey === (selectedItem == null ? void 0 : selectedItem.key);
  if (!content || !isSelected && destroyInactiveTabPanel) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
    Component,
    {
      ref: domRef,
      "data-focus": isFocused,
      "data-focus-visible": isFocusVisible,
      "data-inert": !isSelected ? "true" : void 0,
      inert: (0, import_shared_utils3.getInertValue)(!isSelected),
      ...isSelected && (0, import_shared_utils3.mergeProps)(tabPanelProps, focusProps, otherProps),
      className: (_b = slots.panel) == null ? void 0 : _b.call(slots, { class: tabPanelStyles }),
      "data-slot": "panel",
      children: content
    }
  );
});
TabPanel.displayName = "HeroUI.TabPanel";
var tab_panel_default = TabPanel;

// src/tabs.tsx
var import_jsx_runtime3 = require("react/jsx-runtime");
var Tabs = (0, import_system4.forwardRef)(function Tabs2(props, ref) {
  const {
    Component,
    values,
    state,
    domRef,
    destroyInactiveTabPanel,
    getBaseProps,
    getTabListProps,
    getWrapperProps,
    getTabCursorProps
  } = useTabs({
    ...props,
    ref
  });
  const tabsProps = {
    state,
    listRef: values.listRef,
    slots: values.slots,
    classNames: values.classNames,
    isDisabled: values.isDisabled,
    shouldSelectOnPressUp: values.shouldSelectOnPressUp
  };
  const tabs2 = [...state.collection].map((item) => /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(tab_default, { item, ...tabsProps, ...item.props }, item.key));
  const variant = props == null ? void 0 : props.variant;
  const previousVariant = (0, import_react2.useRef)(void 0);
  const isVertical = props == null ? void 0 : props.isVertical;
  const previousIsVertical = (0, import_react2.useRef)(void 0);
  const placement = props == null ? void 0 : props.placement;
  const previousPlacement = (0, import_react2.useRef)(void 0);
  const cursorRef = (0, import_react2.useRef)(null);
  const selectedItem = state.selectedItem;
  const selectedKey = selectedItem == null ? void 0 : selectedItem.key;
  const getCursorStyles = (0, import_react2.useCallback)(
    (tabRect) => {
      if (variant === "underlined") {
        return {
          left: `${tabRect.left + tabRect.width * 0.1}px`,
          top: `${tabRect.top + tabRect.height - 2}px`,
          width: `${tabRect.width * 0.8}px`,
          height: ""
        };
      }
      return {
        left: `${tabRect.left}px`,
        top: `${tabRect.top}px`,
        width: `${tabRect.width}px`,
        height: `${tabRect.height}px`
      };
    },
    [variant]
  );
  const withAnimationReset = (0, import_react2.useCallback)(
    (callback) => {
      var _a;
      if (variant !== previousVariant.current || isVertical !== previousIsVertical.current || placement !== previousPlacement.current) {
        (_a = cursorRef.current) == null ? void 0 : _a.removeAttribute("data-animated");
      }
      callback();
      previousVariant.current = variant;
      previousIsVertical.current = isVertical;
      previousPlacement.current = placement;
      requestAnimationFrame(() => {
        var _a2, _b;
        (_a2 = cursorRef.current) == null ? void 0 : _a2.setAttribute("data-animated", "true");
        (_b = cursorRef.current) == null ? void 0 : _b.setAttribute("data-initialized", "true");
      });
    },
    [isVertical, variant, placement]
  );
  const updateCursorPosition = (0, import_react2.useCallback)(
    (selectedTab) => {
      if (!cursorRef.current) return;
      const tabRect = {
        width: selectedTab.offsetWidth,
        height: selectedTab.offsetHeight,
        left: selectedTab.offsetLeft,
        top: selectedTab.offsetTop
      };
      const styles = getCursorStyles(tabRect);
      withAnimationReset(() => {
        cursorRef.current.style.left = styles.left;
        cursorRef.current.style.top = styles.top;
        cursorRef.current.style.width = styles.width;
        cursorRef.current.style.height = styles.height;
      });
    },
    [cursorRef.current, getCursorStyles, withAnimationReset]
  );
  const onResize = (0, import_react2.useCallback)(
    (entries) => {
      const contentRect = entries[0].contentRect;
      if (contentRect.width === 0 && contentRect.height === 0) return;
      updateCursorPosition(entries[0].target);
    },
    [updateCursorPosition]
  );
  (0, import_react2.useEffect)(() => {
    var _a;
    const selectedTab = (_a = domRef.current) == null ? void 0 : _a.querySelector(`[data-key="${selectedKey}"]`);
    if (!selectedTab) return;
    const observer = new ResizeObserver(onResize);
    observer.observe(selectedTab);
    return () => observer.disconnect();
  }, [domRef.current, onResize, selectedKey]);
  const renderTabs = (0, import_react2.useMemo)(
    () => /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)(import_jsx_runtime3.Fragment, { children: [
      /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("div", { ...getBaseProps(), children: /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)(Component, { ...getTabListProps(), children: [
        !values.disableAnimation && !values.disableCursorAnimation && selectedKey != null && /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("span", { ...getTabCursorProps(), ref: cursorRef }),
        tabs2
      ] }) }),
      [...state.collection].map((item) => {
        return /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
          tab_panel_default,
          {
            classNames: values.classNames,
            destroyInactiveTabPanel,
            slots: values.slots,
            state: values.state,
            tabKey: item.key
          },
          item.key
        );
      })
    ] }),
    [
      Component,
      getBaseProps,
      getTabListProps,
      getTabCursorProps,
      tabs2,
      selectedKey,
      state.collection,
      values.disableAnimation,
      values.disableCursorAnimation,
      values.classNames,
      values.slots,
      values.state,
      destroyInactiveTabPanel,
      domRef,
      cursorRef,
      variant,
      isVertical
    ]
  );
  if ("placement" in props || "isVertical" in props) {
    return /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("div", { ...getWrapperProps(), children: renderTabs });
  }
  return renderTabs;
});
var tabs_default = Tabs;
