// tailwind-variants/dist/utils.cjs
'use strict';

var chunk2JY7EID6_cjs = require('./chunk-2JY7EID6.cjs');



Object.defineProperty(exports, "cx", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.cx; }
});
Object.defineProperty(exports, "falsyToString", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.falsyToString; }
});
Object.defineProperty(exports, "flat", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.flat; }
});
Object.defineProperty(exports, "flatArray", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.flatArray; }
});
Object.defineProperty(exports, "flatMergeArrays", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.flatMergeArrays; }
});
Object.defineProperty(exports, "isBoolean", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.isBoolean; }
});
Object.defineProperty(exports, "isEmptyObject", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.isEmptyObject; }
});
Object.defineProperty(exports, "isEqual", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.isEqual; }
});
Object.defineProperty(exports, "joinObjects", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.joinObjects; }
});
Object.defineProperty(exports, "mergeObjects", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.mergeObjects; }
});
Object.defineProperty(exports, "removeExtraSpaces", {
  enumerable: true,
  get: function () { return chunk2JY7EID6_cjs.removeExtraSpaces; }
});
