// @heroui/system-rsc/dist/chunk-IIHM6YAA.mjs
import {
  mapPropsVariants
} from "./chunk-YFAKJTDR.mjs";

// src/extend-variants.js
import * as React from "react";
import { tv, cn } from "@heroui/theme";
function getSlots(variants) {
  if (!variants || typeof variants !== "object") return {};
  const acc = /* @__PURE__ */ Object.create(null);
  for (const group of Object.values(variants)) {
    if (!group || typeof group !== "object") continue;
    for (const config of Object.values(group)) {
      if (!config || typeof config !== "object" || Array.isArray(config) || config instanceof String) {
        continue;
      }
      for (const slotName of Object.keys(config)) {
        acc[slotName] = "";
      }
    }
  }
  return acc;
}
function getClassNamesWithProps({
  props = {},
  variants,
  slots,
  defaultVariants,
  compoundVariants,
  hasSlots,
  opts
}) {
  var _a, _b, _c;
  const keys = [];
  if (defaultVariants && typeof defaultVariants === "object") {
    for (const key in defaultVariants) {
      const value = defaultVariants[key];
      const propValue = props == null ? void 0 : props[key];
      if (propValue && propValue !== value) {
        keys.push(key);
      }
    }
  }
  const customTv = tv(
    {
      variants,
      defaultVariants: defaultVariants && typeof defaultVariants === "object" ? (
        // Do not apply default variants when the props variant is different
        Object.keys(defaultVariants).filter((k) => !keys.includes(k)).reduce((o, k) => {
          o[k] = defaultVariants[k];
          return o;
        }, [])
      ) : defaultVariants,
      compoundVariants,
      ...hasSlots && { slots }
    },
    {
      twMerge: (_a = opts.twMerge) != null ? _a : true,
      twMergeConfig: (_b = opts.twMergeConfig) != null ? _b : {}
    }
  );
  const [baseProps, variantProps] = mapPropsVariants(props, customTv.variantKeys, false);
  const newProps = {
    ...defaultVariants,
    ...baseProps,
    className: cn(defaultVariants == null ? void 0 : defaultVariants.className, baseProps.className)
  };
  let classNames = {};
  const result = customTv(variantProps);
  if (!hasSlots) {
    newProps.className = cn(result, newProps.className);
  } else {
    Object.entries(result).forEach(([key, value]) => {
      const slotResult = value();
      if (typeof slotResult === "string") {
        classNames[key] = slotResult;
      }
    });
    Object.entries((_c = props.classNames) != null ? _c : {}).forEach(([key, value]) => {
      classNames[key] = cn(classNames[key], value);
    });
  }
  if (Object.keys(classNames).length !== 0) {
    newProps.classNames = classNames;
  }
  return newProps;
}
function extendVariants(BaseComponent, styles = {}, opts = {}) {
  var _a;
  const { variants, defaultVariants, compoundVariants, slots: directSlots } = styles || {};
  const inheritedVariants = (_a = BaseComponent.__variants) != null ? _a : {};
  const mergedVariants = {
    ...inheritedVariants,
    ...variants
  };
  const inferredSlots = getSlots(mergedVariants);
  const slots = directSlots ? { ...inferredSlots, ...directSlots } : inferredSlots;
  const hasSlots = typeof slots === "object" && Object.keys(slots).length !== 0;
  const ForwardedComponent = React.forwardRef((originalProps = {}, ref) => {
    const newProps = React.useMemo(
      () => getClassNamesWithProps(
        {
          slots,
          variants: mergedVariants,
          compoundVariants,
          props: originalProps,
          defaultVariants,
          hasSlots,
          opts
        },
        [originalProps]
      )
    );
    return React.createElement(BaseComponent, { ...originalProps, ...newProps, ref });
  });
  if (BaseComponent.getCollectionNode) {
    ForwardedComponent.getCollectionNode = (itemProps) => {
      const newProps = getClassNamesWithProps({
        slots,
        variants: mergedVariants,
        compoundVariants,
        props: itemProps,
        defaultVariants,
        hasSlots,
        opts
      });
      return BaseComponent.getCollectionNode({ ...itemProps, ...newProps });
    };
  }
  ForwardedComponent.displayName = `Extended(${BaseComponent.displayName || BaseComponent.name})`;
  ForwardedComponent.__variants = mergedVariants;
  return ForwardedComponent;
}

export {
  extendVariants
};
